packages/composite-checkout/src/components/checkout-provider.tsx

```diff
@@ -195,8 +195,17 @@ function useCallEventCallbacks( {
 	paymentMethodId: string | null;
 	transactionLastResponse: PaymentProcessorResponseData;
 } ): void {
+	// Ensure that onPaymentComplete is only called once per checkout so that the URL
+	// retrieved from the signup destination cookie will be used correctly.
+	const didOnPaymentComplete = useRef( false );
```

Comment:
While this seems reasonable, I have a few concerns. Specifically, it seems like there might be an issue with the dependencies of the `useEffect` below changing unexpectedly, and this flag only hides that bug. Are you able to reproduce a situation where `onPaymentComplete` is being called more than once? And if so, which of the dependencies is causing that?

If we add this flag, it's very final and it limits the behavior of the component in ways that might not be obvious to us now and could result in confusing behavior later.

It seems to me that none of the dependencies should change after a form has been completed unless, for example, `onPaymentComplete` is not memoized, but that might cause other problems. If the issue is that `onPaymentComplete` is not memoized, we may be able to resolve that in a different way.

If we do change this behavior, we do we also need make a similar change for `onPaymentRedirect` below?

Reply:
I'm working on an alternative to this change in https://github.com/Automattic/wp-calypso/pull/59499